presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRedundantSortColumnsRemoval.java

```diff
@@ -0,0 +1,124 @@
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY custkey, orderkey, sum(totalprice), min(orderdate) LIMIT 10")
+                .matches(topNMatchWith(ImmutableList.of(sort("CUSTKEY", ASCENDING, LAST), sort("ORDERKEY", ASCENDING, LAST))));
+
+        // No impact due to sort direction
+        tester().assertThat(ImmutableSet.of(new MergeLimitWithSort(), new RemoveRedundantTopNColumns()), logicalPropertiesProvider)
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY custkey DESC, orderkey ASC, sum(totalprice), min(orderdate) LIMIT 10")
+                .matches(topNMatchWith(ImmutableList.of(sort("CUSTKEY", DESCENDING, LAST), sort("ORDERKEY", ASCENDING, LAST))));
+
+        // Negative test - No prefix matches the grouping set, so TopN columns are not pruned
+        tester().assertThat(ImmutableSet.of(new MergeLimitWithSort(), new RemoveRedundantTopNColumns()), logicalPropertiesProvider)
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY orderkey, sum(totalprice), custkey, min(orderdate) LIMIT 10")
+                .doesNotMatch(topNMatchWith(ImmutableList.of(sort("ORDERKEY", ASCENDING, LAST), sort("CUSTKEY", ASCENDING, LAST))));
+    }
+
+    @Test
+    public void testRemoveRedundantColumnsFromSort()
+    {
+        // OrderBy prefix matches GroupBy columns clause exactly
+        tester().assertThat(ImmutableSet.of(new RemoveRedundantSortColumns()), logicalPropertiesProvider)
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY orderkey, custkey, sum(totalprice), min(orderdate)")
+                .matches(sortWith(ImmutableList.of(sort("ORDERKEY", ASCENDING, LAST), sort("CUSTKEY", ASCENDING, LAST))));
+
+        // Flipped order matches too since the Grouping set remains the same
+        tester().assertThat(ImmutableSet.of(new RemoveRedundantSortColumns()), logicalPropertiesProvider)
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY custkey, orderkey, sum(totalprice), min(orderdate)")
+                .matches(sortWith(ImmutableList.of(sort("CUSTKEY", ASCENDING, LAST), sort("ORDERKEY", ASCENDING, LAST))));
+
+        // No impact due to sort direction
+        tester().assertThat(ImmutableSet.of(new RemoveRedundantSortColumns()), logicalPropertiesProvider)
+                .on("SELECT orderkey, custkey, sum(totalprice), min(orderdate) FROM orders GROUP BY orderkey, custkey " +
+                        "ORDER BY custkey DESC, orderkey ASC, sum(totalprice), min(orderdate)")
```

Comment:
Would this rule run if both `ORDER BY` fields are `DESC`? Is it worth adding a case for that?

Reply:
Yes it will run. IMO the extra case is not worth it